draft-irtf-cfrg-opaque.md

```diff
@@ -398,145 +398,331 @@ This specification defines one particular AKE based on 3DH; see {{instantiations
+The function CreateCleartextCredentials implements this logic:
+~~~
+CreateCleartextCredentials(server_public_key, client_public_key, server_identity, client_identity)
 
-                               response
-                      <-------------------------
+Input:
+- server_public_key, The encoded client public key for the AKE protocol
+- client_public_key, The encoded client public key for the AKE protocol
+- server_identity, The encoded server identity
+- client_identity, The encoded client identity
 
- (record, export_key) = FinalizeRequest(password, creds, blind, response)
+Output:
+- cleartext_credentials, a CleartextCredentials structure
 
-                                record
-                      ------------------------->
+Steps:
+1. if server_public_key == nil || client_public_key == nil
+2.  raise NoPubKeyError
+3. if server_identity == nil
+4.  server_identity = server_public_key
+5. if client_identity == nil
+6.  client_identity = client_public_key
+8. Create CleartextCredentials cleartext_credentials with (server_public_key, client_public_key, server_identity, client_identity)
+9. Output cleartext_credentials
 ~~~
 
-{{registration-functions}} describes details of the functions referenced above.
+During protocol execution, these values can be stored in an implementation-specific `Credentials` object with names matching the values. For example, `Credentials.server_public_key` is the server's public key.
 
-Both client and server MUST validate the other party's public key before use.
-See {{validation}} for more details.
+## Envelope Structure {#envelope-structure}
 
-Upon completion, S stores C's credentials for later use. See {{credential-file}}
-for a recommended storage format.
+A client `Envelope` is constructed independently from the `EnvelopeMode`, but offers an `InnerEnvelope` entry that modes can use as they see fit.
+`Envelope` is constructed as follows:
+~~~
+struct {
+ EnvelopeMode mode;
+ opaque nonce[Nn];
+ opaque auth_tag[Nh];
+ InnerEnvelope inner_env;
+} Envelope;
+~~~
 
-## Credential Storage {#credential-storage}
+mode : The `EnvelopeMode` used for the `Envelope`.
 
-OPAQUE makes use of a structure `Envelope` to store client credentials.
-The `Envelope` structure embeds the following types of credentials:
+nonce : A unique nonce of length `Nn` used to protect this Envelope.
 
-- client_private_key: The encoded client private key for the AKE protocol.
-- server_public_key: The encoded server public key for the AKE protocol.
-- client_identity: The client identity. This is an application-specific value, e.g., an e-mail
-  address or normal account name.
-- server_identity: The server identity. This is typically a domain name, e.g., example.com.
-  See {{identities}} for information about this identity.
+auth_tag : Authentication tag protecting the contents of the envelope, covering `InnerEnvelope` and `CleartextCredentials`.
```

Comment:
This also includes `mode` and `nonce`, right?

Reply:
Yes, it does! Well spotted.